File: backend/scripts/generateStrongsCSV.py
import csv
import re
import os

# Paths
DATA_DIR = os.path.join(os.path.dirname(__file__), '..', 'data')
HEBREW_FILE = os.path.join(DATA_DIR, 'HebrewStrongs.csv')
GREEK_FILE = os.path.join(DATA_DIR, 'GreekStrongs.csv')
TELUGU_FILE = os.path.join(DATA_DIR, 'TeluguHindiStrongs.csv')
OUTPUT_FILE = os.path.join(DATA_DIR, 'StrongsConcordance-EnglishTelugu.csv')

def clean_gloss(gloss):
    """Extract clean English meaning from gloss field"""
    if not gloss:
        return ''
    # Remove the pronunciation guide at start
    gloss = re.sub(r'^[^,]+,\s*', '', gloss, count=1)
    # Get first definition line
    lines = gloss.strip().split('\n')
    meanings = []
    for line in lines:
        line = line.strip()
        if line.startswith('1.') or line.startswith('2.') or line.startswith('3.'):
            # Extract the meaning
            meaning = re.sub(r'^\d+\.\s*', '', line)
            meaning = re.sub(r'\{[^}]*\}', '', meaning)  # Remove bracketed notes
            meaning = re.sub(r'\[[^\]]*\]', '', meaning)  # Remove square brackets
            meaning = meaning.strip()
            if meaning:
                meanings.append(meaning)
    if meanings:
        return '; '.join(meanings[:3])  # Return first 3 meanings
    # Fallback: return first line cleaned
    first_line = lines[0] if lines else ''
    first_line = re.sub(r'\{[^}]*\}', '', first_line)
    first_line = re.sub(r'\[[^\]]*\]', '', first_line)
    return first_line.strip()

def load_telugu_meanings():
    """Load Telugu meanings from TeluguHindiStrongs.csv"""
    telugu = {}
    try:
        with open(TELUGU_FILE, 'r', encoding='utf-8') as f:
            reader = csv.DictReader(f)
            for row in reader:
                strongs = row.get('strongs_number', '').strip()
                telugu_meaning = row.get('telugu_meaning', '').strip()
                if strongs and telugu_meaning:
                    # Normalize strongs number (H0001 -> H1)
                    if strongs.startswith('H') or strongs.startswith('G'):
                        num = int(re.sub(r'[HGA]', '', strongs))
                        normalized = f"{strongs[0]}{num}"
                        telugu[normalized] = telugu_meaning
    except Exception as e:
        print(f"Error loading Telugu file: {e}")
    return telugu

def load_hebrew_words():
    """Load Hebrew words from HebrewStrongs.csv"""
    words = []
    try:
        with open(HEBREW_FILE, 'r', encoding='utf-8-sig') as f:  # utf-8-sig handles BOM
            reader = csv.DictReader(f)
            for row in reader:
                strongs_num = row.get('strongs_number', '').strip()
                if not strongs_num:
                    continue
                # strongs_num is just a number like "1", "2", etc.
                try:
                    num = int(strongs_num)
                except:
                    continue
                    
                word = row.get('word', '').strip()
                gloss = row.get('gloss', '').strip()
                language = row.get('language', 'H').strip()
                pos = row.get('part_of_speech', '').strip()
                occurrences = row.get('occurrences', '').strip()
                
                # Format strongs number with H or A prefix
                strongs_formatted = f"{language}{num}"
                
                words.append({
                    'strongs_number': strongs_formatted,
                    'original_word': word,
                    'testament': 'Old Testament',
                    'language': 'Hebrew' if language == 'H' else 'Aramaic',
                    'part_of_speech': pos,
                    'english_meaning': clean_gloss(gloss),
                    'occurrences': occurrences
                })
    except Exception as e:
        print(f"Error loading Hebrew file: {e}")
        import traceback
        traceback.print_exc()
    return words

def load_greek_words():
    """Load Greek words from GreekStrongs.csv (tab-separated)"""
    words = []
    try:
        with open(GREEK_FILE, 'r', encoding='utf-8') as f:
            reader = csv.reader(f, delimiter='\t')
            headers = next(reader)  # Skip header
            for row in reader:
                if len(row) < 2:
                    continue
                strongs_num = row[0].strip()
                lemma = row[1].strip() if len(row) > 1 else ''
                origin = row[2].strip() if len(row) > 2 else ''
                
                if not strongs_num.startswith('G'):
                    continue
                
                # For Greek, we'll use origin/root info as meaning hint
                words.append({
                    'strongs_number': strongs_num,
                    'original_word': lemma,
                    'testament': 'New Testament',
                    'language': 'Greek',
                    'part_of_speech': '',
                    'english_meaning': f"From {origin}" if origin else lemma,
                    'occurrences': ''
                })
    except Exception as e:
        print(f"Error loading Greek file: {e}")
    return words

def main():
    print("Loading Telugu meanings...")
    telugu_meanings = load_telugu_meanings()
    print(f"Loaded {len(telugu_meanings)} Telugu meanings")
    
    print("Loading Hebrew words...")
    hebrew_words = load_hebrew_words()
    print(f"Loaded {len(hebrew_words)} Hebrew/Aramaic words")
    
    print("Loading Greek words...")
    greek_words = load_greek_words()
    print(f"Loaded {len(greek_words)} Greek words")
    
    # Combine all words
    all_words = hebrew_words + greek_words
    
    # Add Telugu meanings
    for word in all_words:
        strongs = word['strongs_number']
        word['telugu_meaning'] = telugu_meanings.get(strongs, '')
    
    # Write output CSV
    print(f"Writing output to {OUTPUT_FILE}...")
    with open(OUTPUT_FILE, 'w', encoding='utf-8', newline='') as f:
        fieldnames = [
            'strongs_number',
            'original_word', 
            'testament',
            'language',
            'part_of_speech',
            'english_meaning',
            'telugu_meaning',
            'occurrences'
        ]
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        writer.writerows(all_words)
    
    print(f"Done! Created {OUTPUT_FILE}")
    print(f"Total entries: {len(all_words)}")
    print(f"Entries with Telugu: {sum(1 for w in all_words if w['telugu_meaning'])}")

if __name__ == '__main__':
    main()

File: backend/src/scripts/generateStrongsCSV.js
/**
 * Script to generate a comprehensive CSV file with Hebrew (OT) and Greek (NT) words
 * with English and Telugu meanings
 */

const fs = require('fs');
const path = require('path');

// Read and parse Hebrew Strong's
function parseHebrewStrongs() {
    const filePath = path.join(__dirname, '../../data/HebrewStrongs.csv');
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');

    const entries = [];
    let currentRecord = [];
    let inQuotes = false;
    let isFirstLine = true;

    for (const line of lines) {
        if (isFirstLine) {
            isFirstLine = false;
            continue; // Skip header
        }

        const quoteCount = (line.match(/"/g) || []).length;

        if (!inQuotes) {
            currentRecord = [line];
            if (quoteCount % 2 !== 0) {
                inQuotes = true;
            } else {
                const entry = parseHebrewRecord(currentRecord.join('\n'));
                if (entry) entries.push(entry);
            }
        } else {
            currentRecord.push(line);
            if (quoteCount % 2 !== 0) {
                inQuotes = false;
                const entry = parseHebrewRecord(currentRecord.join('\n'));
                if (entry) entries.push(entry);
            }
        }
    }

    return entries;
}

function parseHebrewRecord(record) {
    if (!record.trim()) return null;

    const parts = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < record.length; i++) {
        const char = record[i];

        if (char === '"' && !inQuotes) {
            inQuotes = true;
        } else if (char === '"' && inQuotes) {
            if (record[i + 1] === '"') {
                current += '"';
                i++;
            } else {
                inQuotes = false;
            }
        } else if (char === ',' && !inQuotes) {
            parts.push(current.trim());
            current = '';
        } else {
            current += char;
        }
    }
    parts.push(current.trim());

    if (parts.length >= 3 && parts[0]) {
        // Extract just the main meaning from the gloss (first line, simplified)
        let gloss = parts[2] || '';
        // Clean up - get just the main meaning
        const glossMatch = gloss.match(/n-[mf]?\.\s*\d+\.\s*([^\n{[]+)/);
        let simpleMeaning = glossMatch ? glossMatch[1].trim() : gloss.split('\n')[0].replace(/["{}\[\]]/g, '').trim();

        // Further simplify - take first meaningful part
        simpleMeaning = simpleMeaning.split(/[,;]/)[0].trim();
        if (simpleMeaning.length > 100) {
            simpleMeaning = simpleMeaning.substring(0, 100) + '...';
        }

        return {
            strongsNumber: `H${parts[0].padStart(4, '0')}`,
            word: parts[1] || '',
            englishMeaning: simpleMeaning || parts[2]?.split('\n')[0] || '',
            language: 'Hebrew',
            testament: 'Old Testament'
        };
    }
    return null;
}

// Read and parse Greek Strong's (TSV format)
function parseGreekStrongs() {
    const filePath = path.join(__dirname, '../../data/GreekStrongs.csv');
    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');

    const entries = [];
    let isFirstLine = true;

    for (const line of lines) {
        if (isFirstLine) {
            isFirstLine = false;
            continue; // Skip header
        }

        if (!line.trim()) continue;

        const parts = line.split('\t');
        if (parts.length >= 2 && parts[0]) {
            const strongsNumber = parts[0].trim();
            const lemma = parts[1]?.trim() || '';
            const rootLemma = parts[4]?.trim() || '';

            entries.push({
                strongsNumber: strongsNumber,
                word: lemma,
                englishMeaning: rootLemma || lemma, // Use root lemma as hint
                language: 'Greek',
                testament: 'New Testament'
            });
        }
    }

    return entries;
}

// Read Telugu translations
function parseTeluguStrongs() {
    const filePath = path.join(__dirname, '../../data/TeluguHindiStrongs.csv');

    if (!fs.existsSync(filePath)) {
        return new Map();
    }

    const content = fs.readFileSync(filePath, 'utf8');
    const lines = content.split('\n');

    const teluguMap = new Map();
    let isFirstLine = true;

    for (const line of lines) {
        if (isFirstLine) {
            isFirstLine = false;
            continue;
        }

        if (!line.trim()) continue;

        const parts = line.split(',');
        if (parts.length >= 3) {
            const strongsNum = parts[0].trim().toUpperCase();
            const teluguWord = parts[1]?.trim() || '';
            const teluguMeaning = parts[2]?.trim() || '';

            teluguMap.set(strongsNum, {
                word: teluguWord,
                meaning: teluguMeaning
            });
        }
    }

    return teluguMap;
}

// Main function
function generateCSV() {
    console.log('📖 Generating Hebrew and Greek Strong\'s CSV with Telugu meanings...');

    // Parse all data
    console.log('  Loading Hebrew Strong\'s...');
    const hebrewEntries = parseHebrewStrongs();
    console.log(`    Found ${hebrewEntries.length} Hebrew entries`);

    console.log('  Loading Greek Strong\'s...');
    const greekEntries = parseGreekStrongs();
    console.log(`    Found ${greekEntries.length} Greek entries`);

    console.log('  Loading Telugu translations...');
    const teluguMap = parseTeluguStrongs();
    console.log(`    Found ${teluguMap.size} Telugu entries`);

    // Combine all entries
    const allEntries = [...hebrewEntries, ...greekEntries];

    // Add Telugu meanings
    for (const entry of allEntries) {
        const telugu = teluguMap.get(entry.strongsNumber);
        if (telugu) {
            entry.teluguWord = telugu.word;
            entry.teluguMeaning = telugu.meaning;
        } else {
            entry.teluguWord = '';
            entry.teluguMeaning = '';
        }
    }

    // Generate CSV content
    const csvHeader = 'Strong\'s Number,Original Word,English Meaning,Telugu Word,Telugu Meaning,Language,Testament';
    const csvRows = allEntries.map(entry => {
        // Escape fields that might contain commas or quotes
        const escape = (str) => {
            if (!str) return '';
            if (str.includes(',') || str.includes('"') || str.includes('\n')) {
                return `"${str.replace(/"/g, '""')}"`;
            }
            return str;
        };

        return [
            escape(entry.strongsNumber),
            escape(entry.word),
            escape(entry.englishMeaning),
            escape(entry.teluguWord),
            escape(entry.teluguMeaning),
            escape(entry.language),
            escape(entry.testament)
        ].join(',');
    });

    const csvContent = [csvHeader, ...csvRows].join('\n');

    // Write to file
    const outputPath = path.join(__dirname, '../../data/StrongsWithTelugu.csv');
    fs.writeFileSync(outputPath, csvContent, 'utf8');

    console.log(`\n✅ CSV generated successfully!`);
    console.log(`   Output: ${outputPath}`);
    console.log(`   Total entries: ${allEntries.length}`);
    console.log(`   - Hebrew (OT): ${hebrewEntries.length}`);
    console.log(`   - Greek (NT): ${greekEntries.length}`);
    console.log(`   - With Telugu: ${allEntries.filter(e => e.teluguMeaning).length}`);
}

generateCSV();
